rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,364 @@
+while letting users insert their own custom input logic.
+
+
+### What to focus first?
+
+Gamepad input assumes an initially focused element
+for navigating "move in direction" or "press action".
+At the very beginning of execution, there are no focused element,
+so we need fallbacks.
+By default, it is any `Focusable` when none are focused yet.
+But the user can spawn a `Focusable` as `dormant`
+and the algorithm will prefer it when no focused nodes exist yet.
+
+```rust
+self.focusables
+    .iter()
+    // The focused focusable.
+    .find_map(|(e, focus)| (focus.state == Focused).then(|| e))
+    // Dormant focusable within the root menu (if it exists)
+    .or_else(root_dormant)
+    // Any dormant focusable
+    .or_else(any_dormant)
+    // Any focusable
+    .or_else(fallback)
+```
+
+
+### Menu navigation handling
+
+The public API has `NavMenu`, but we use internally `TreeMenu`,
+this prevents end users from breaking assumptions about the menu trees.
+
+We define the `TreeMenu` component.
+All `Focusable` direct or indirect children of a `TreeMenu` are part of that menu.
+A `TreeMenu` has a `Focusable` focus parent,
+when that `Focusable` is focused and the `Action` request is received,
+the focus changes to the `TreeMenu`
+
+In short, the navigation tree is built on two elements:
+- The `reachable_from` that a `NavMenu` was spawned with
+- The `Focusable` within the hierarchy of a `NavMenu`
+
+This is a tree in the ECS.
+
+In the following screenshot, `Focusable`s are represented as circles, and menus
+as rectangles.
+
+![A screenshot of a RPG-style menu with the navigation tree overlayed](https://user-images.githubusercontent.com/26321040/141671969-ea4a461d-0949-4b98-b060-c492d7b896bd.png)
+
+A `Focusable` used to go from the root menu
+to the menu in which the currently focused entity is is _active_.
+The current focus position is represented as a trail of breadcrumb active
+focusables and the currently focused focusable.
+
+This is reflected into the `NavEvent::FocusChanged` event.
+When going from a focusable to another,
+the navigation system emits a `FocusChanged` event,
+the event contains two lists of entities:
+- The list of entities that after the navigation event are no more focused
+  or active
+- The list of entities that are newly active or focused
+
+For the simplest case of navigating from one focusable to another
+in the same menu, the two list have a single element,
+so it is capital that it is easy to access those elements.
+This is why we use a `NonEmpty`, from the `non_empty_vec` crate.
+A `NonEmpty<T>` has a `first` method that returns a `T`,
+which in our case would be the last and new focused elements.
+I chose this crate over similar ones,
+because the code is dead simple and easy to vet.
+
+#### Going back to a previous menus
+
+`Focusable`s have a `dormant` state that is set when they go from
+active/focused to not active.
+This is a form of memory that allows focus to go back to the last focused
```

Comment:
This is a good feature, but I'm not sure this is the clearest implementation? I think most of my problem here is that the "dormant" name isn't intuitive to me.

Reply:
I spent a lot of time trying to come up with a good name here.

The difficulty I had was to **not** make it sound like it is the opposite of active.
[Here is the whole enum list for reference](https://docs.rs/bevy-ui-navigation/latest/bevy_ui_navigation/enum.FocusState.html)

Reply:
`PreviouslyActive`?

Reply:
I went with `Prioritized`. It doesn't tie much with the fact that it was previously active, but in practice, the variant is only used to chose which focusable to focus first in a menu, so the name makes sense.